10/calculator_test.go

```diff
@@ -0,0 +1,101 @@
+package calculator_test
+
+import (
+	"testing"
+	"calculator"
+)
+
+func TestAdd(t *testing.T) {
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 2},
+		{a: 2, b: 2, want: 4},
+		{a: 6, b: 7, want: 13},
+		{a: -6, b: -7, want: -13},
+		{a: 6, b: -7, want: -1},
+		{a: 0, b: -7, want: -7},
+		{a: 0, b: 0, want: 0},
+		{a: -1, b: 0, want: -1},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Add(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Add(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 0},
+		{a: 2, b: 1, want: 1},
+		{a: 6, b: 7, want: -1},
+		{a: 7, b: 6, want: 1},
+		{a: -6, b: -7, want: 1},
+		{a: 6, b: -7, want: 13},
+		{a: 0, b: -7, want: 7},
+		{a: 0, b: 0, want: 0},
+		{a: -1, b: 0, want: -1},
+		{a: -1, b: 7, want: -8},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Subtract(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Subtract(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestMultiply(t *testing.T) {
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 1},
+		{a: 2, b: 1, want: 2},
+		{a: 6, b: 7, want: 42},
+		{a: -6, b: -7, want: 42},
+		{a: 6, b: -7, want: -42},
+		{a: 0, b: -7, want: 0},
+		{a: 0, b: 0, want: 0},
+		{a: -1, b: 0, want: 0},
+		{a: -1, b: 7, want: -7},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Multiply(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Multiply(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestDivide(t *testing.T) {
+	testCases := []struct {
+		a, b        int
+		want        int
+		errExpected bool
+	}{
+		{a: 1, b: 1, want: 1, errExpected: false},
+		{a: 2, b: 1, want: 2, errExpected: false},
+		{a: 6, b: -7, want: 0, errExpected: false},
+		{a: -1, b: 7, want: 0, errExpected: false},
+		{a: 10, b: 0, want: 0, errExpected: true},
+	}
+	for _, testCase := range testCases {
+		got, _ := calculator.Divide(testCase.a, testCase.b)
+		if testCase.want != got {
```

Comment:
Since in the error case, the data value is to be ignored, we should check the error value _first_, shouldn't we? Only if an error is not expected, and not received, is it legitimate to check the data value against the test case.

Reply:
thanks @bitfield yep it makes sense. I was thinking for something like the following?

```
if (testCase.errExpected == false && err == nil) && testCase.want != got {
      t.Errorf("Divide(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
}
```